live-build/hooks/601-create-dpkg-yaml.binary

```diff
@@ -0,0 +1,30 @@
+#!/bin/sh
+#
+# create dpkg.yaml used for OSS compliance
+
+set -ex
+
+echo "I: Creating dpkg.yaml needed for OSS compliance"
+
+PREFIX=binary/boot/filesystem.dir
+
+(cd $PREFIX
+	{
+	# fill in ppa information in yaml file
+	  printf 'package-repositories:\n'
+	  find etc/apt/ -name \*.list | while IFS= read -r APT; do
+	    grep -o "^deb http://ppa.launchpad.net/[a-z0-9\.\+\-]\+/[a-z0-9\.\+\-]\+/[a-z0-9\.\+\-]\+" "$APT" | while read -r ENTRY ; do
+	      USER=$(echo "$ENTRY" | cut -d/ -f4)
+	      PPA=$(echo "$ENTRY" | cut -d/ -f5)
+	      DISTRO=$(echo "$ENTRY" | cut -d/ -f6)
+	      printf -- '- type: apt\n'
+	      echo '  ppa: '"$USER/$DISTRO/$PPA"
+	    done
+	  done
+	  
+	# fill in yaml section with all installed packages
+	  printf 'packages:\n'
+	  tail -n +6 usr/share/snappy/dpkg.list | awk '{printf "- "}{printf $2}{printf "="}{print $3}'
```

Comment:
I would suggest to use something like `grep ^ii usr/share/snappy/dpkg.list | awk '{printf "- " $2 "=" $3 "\n"}'` here. This avoids hard-coding the header and also avoids strangeness from dpkg creeping in. The awk line is also slightly more compat this way. I guess one could also use an "if" in the awk line directly but I have no strong opion here.

Reply:
It might also be worth considering generating this separately using something based on `chroot . dpkg-query -W` (possibly with a `--showformat` option as well) rather than based on `dpkg.list`, since the output of `dpkg-query -W` is much more easily parseable than the output of `dpkg -l`.

Reply:
... though if you're doing that then you'll need to move this into `600-no-debian.binary`, since that hook purges several packages just after generating `dpkg.list`.

Reply:
It looks like we won't be able to rely on "ii" in grep as the desired action and package status could be different from package to package and our new generate-yaml hook will miss those when creating the yaml file (one concrete example we currently have with your suggestion would be the dirmngr package in dpkg.list (https://pastebin.canonical.com/p/3ZtCmTcQfD/) will be missed when creating dpkgl.yaml). 

Considering that there are 5 possible desired actions and 8 package statues that dpkg might generate enaching that grep statement and maintaining it rapidly becomes unfeasible. 

However I do agree with your point on not hardcoding the header with a "tail n+" so I'm proposing we actually compute where the packages section begins instead of hardcoding it.

Reply:
I don't see the dirmngr in the paste, am I missing something?

Reply:
Hi Michael, dirmngr is an example where `grep ^ii usr/share/snappy/dpkg.list` would fail to append it to the yaml file, it wouldn't be in the file. The current commit uses a non-harcoded strategy where we actually compute where the packages section begins instead of hardcoding it.

Although this works I will make another update to all 3 snaps that does not rely on calculating the line where the actual packages being nor on grep statements. I will use Colin's suggestion above where we rely on output from `dpkg-query` as that will be indeed the most robust one. Will update with the next commit.

Reply:
And done - generation of dpkg.yaml moved to `600-no-debian.binary` and is now relying on `chroot $PREFIX dpkg-query -W`. Generated yaml file: https://pastebin.canonical.com/p/5hvSdNHd95/

Reply:
Thanks, I really like the suggestion to use dpkg-query